src/AIS.cpp

```diff
@@ -129,3 +136,147 @@ ais_base::VoyageInformation AIS::getVoyageInformation(ais::message_05 const& mes
+}
+
+/**
+ * @brief Converts a GPS position into UTM coordinates
+ *
+ * @param position The GPS position
+ * @param utm_converter The UTM converter
+ *
+ * @return A RigidBodyState in UTM coordinates
+ */
+base::samples::RigidBodyState convertGPSToUTM(ais_base::Position const& position,
+    gps_base::UTMConverter const& utm_converter)
+{
+    gps_base::Solution sensor2world_solution;
+    sensor2world_solution.latitude = position.latitude.getDeg();
+    sensor2world_solution.longitude = position.longitude.getDeg();
+
+    base::samples::RigidBodyState sensor2world_UTM;
+    sensor2world_UTM.position =
+        utm_converter.convertToUTM(sensor2world_solution).position;
+    sensor2world_UTM.time = base::Time::now();
+
+    return sensor2world_UTM;
+}
+
+/**
+ * @brief Converts a UTM position back to GPS coordinates in the world frame
+ *
+ * @param sensor2world_UTM The sensor position in UTM coordinates
+ * @param sensor2vessel_in_world_pos The sensor position in the world frame
+ * @param utm_converter The UTM converter
+ *
+ * @return A pair containing the latitude and longitude
+ */
+std::pair<base::Angle, base::Angle> convertUTMToGPSInWorldFrame(
+    base::samples::RigidBodyState const& sensor2world_UTM,
+    base::Vector3d const& sensor2vessel_in_world_pos,
+    gps_base::UTMConverter const& utm_converter)
+{
+    base::samples::RigidBodyState vessel2world_UTM;
+    vessel2world_UTM.position = sensor2world_UTM.position + sensor2vessel_in_world_pos;
+
+    gps_base::Solution vessel2world_GPS;
+    vessel2world_GPS = utm_converter.convertUTMToGPS(vessel2world_UTM);
+
+    return {base::Angle::fromDeg(vessel2world_GPS.latitude),
+        base::Angle::fromDeg(vessel2world_GPS.longitude)};
+}
+
+ais_base::Position AIS::applyPositionCorrection(
+    std::optional<ais_base::Position> position,
+    base::Vector3d const& sensor2vessel_pos,
+    gps_base::UTMConverter const& utm_converter)
+{
+    if (!position.has_value()) {
+        std::string error_msg = "Position data is unavailable.";
+        LOG_ERROR_S << error_msg;
+        throw std::invalid_argument(error_msg);
+    }
+
+    auto position_value = position.value();
+
+    if (std::isnan(position_value.yaw.getRad()) &&
+        std::isnan(position_value.course_over_ground.getRad())) {
+        LOG_DEBUG_S << "Position can't be corrected because both 'yaw' "
+                       "and 'course_over_ground' values are missing."
+                    << std::endl;
+        return position_value;
+    }
+
+    auto [vessel2world_ori, status] = vesselToWorldOrientation(position_value.yaw,
+        position_value.course_over_ground,
+        position_value.speed_over_ground);
+
+    if (status == ais_base::PositionCorrectionStatus::POSITION_RAW) {
+        LOG_DEBUG_S << "Position can't be corrected because 'yaw' value is missing and "
+                       "'speed_over_ground' is below the threshold."
+                    << std::endl;
+        return position_value;
```

Comment:
You still should propagate the status. Don't assume the parameter has the right status.